File: solutions/0/0/5/3/p0053_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 53: Combinatoric Selections [Level 1]. """
from __future__ import annotations

from solver.runners import runner


@runner.main
def solve(*args: str) -> str:
    """Count C(n, r) above the threshold by walking each Pascal row with the recurrence
    C(n, r+1) = C(n, r) * (n - r) / (r + 1). Rows are unimodal and symmetric, so the first r that
    exceeds the threshold makes entries r..n-r qualify; add n - 2*r + 1 and stop. O(max_n^2) worst
    case, far less with the early exit."""
    max_n = runner.parse_int(args[0])
    threshold = runner.parse_int(args[1])

    count = 0
    for n in range(1, max_n + 1):
        c = 1
        for r in range(0, n // 2 + 1):
            if c > threshold:
                count += n - 2 * r + 1
                break
            else:
                c = c * (n - r) // (r + 1)
    return str(count)


if __name__ == "__main__":
    raise SystemExit(solve())

File: solutions/0/0/5/3/p0053_s0.c
/* Solution to Euler Problem 53: Combinatoric Selections. */
#include "runner.h"

const char *solve(int argc, char *argv[]) {
    /*
     * Count binomial coefficients C(n, r) above the threshold by walking each Pascal row with the
     * recurrence C(n, r+1) = C(n, r) * (n - r) / (r + 1). Each row is unimodal and symmetric, so the
     * first r whose coefficient exceeds the threshold makes every entry from r to n-r qualify; add
     * n - 2*r + 1 and stop. O(max_n^2) worst case, far less with the early exit. The long long
     * accumulator carries intermediate coefficients without overflow for n up to 100.
     */
    static char _answer[32];
    int max_n = parse_int(argv[1]);
    long long threshold = parse_int(argv[2]);

    long long count = 0;
    for (int n = 1; n <= max_n; n++) {
        long long c = 1;
        for (int r = 0; r <= n / 2; r++) {
            if (c > threshold) {
                count += n - 2 * r + 1;
                break;
            } else {
                c = c * (n - r) / (r + 1);
            }
        }
    }
    { snprintf(_answer, sizeof _answer, "%lld", (long long)(count)); return _answer; }
}